### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Jrmcuervo**
In this project we :

- Added integration specs using Capybara and Rspec
- Tested views: users/index, users/show, posts/index, and posts/view
- Checked if correct content is displayed on each page
- Checked if all links on the pages are working
- Fixed n+1 problem in the posts/index view

**Before**

before 1

before 2

**After**
after

### CODE REVIEW

File: spec/views/users/users_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      end
+
+      within page.all('ul li.userNameWrapper').last do
+        expect(page).to have_css('img[alt="Profile P."]')
+        expect(page).to have_link(user2.name, href: user_path(user2), class: 'userNameAnchor')
+        expect(page).to have_content("Number of posts: #{user2.posts.count}")
+      end
+    end
+  end
+end
```
Comment (Qoosim) [inline spec/views/users/users_index_spec.rb]: - [ ] Great job team. There're few tests you still need to write for your user index page. Kindly check below screenshot to see the remaining tests.

### Project requirements.
user-index

File: spec/views/users/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        else
+          expect(page).to have_css('p', text: "#{post.text.split.take(18).join(' ')}...")
+        end
+        expect(page).to have_css('.count', text: "Comments: #{post.comments.count}")
+        expect(page).to have_css('.count', text: "Likes: #{post.likes.count}")
+      end
+    end
+
+    expect(page).to have_css('.button #all-posts-link', text: 'See all posts')
+  end
```
Comment (Qoosim) [inline spec/views/users/users_show_spec.rb:36]: - [ ] Some tests are yet to be written here too. Kindly check the underlined points and write their respective tests for them.

### Project requirement
user-show

File: spec/views/posts/show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+equire 'rails_helper'
```
Comment (Qoosim) [inline spec/views/posts/show_spec.rb]: - [ ] There's a spelling error here. Kindly change that tp `require` instead of `equire`

**Review by Qoosim — CHANGES_REQUESTED** [review]
#  *Status: Required Changes* :recycle:

Hi @team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
-  Some requirements fufilled. :heavy_check_mark: 
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 




## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

- [ ] Kindly look into some previous tests which are failing at this moment

error-1

error-2


_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/users/users_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        expect(page).to have_css('h3', text: "Post ##{index + 1}")
+        if post.text.split.length <= 18
+          expect(page).to have_css('p', text: post.text)
+        else
+          expect(page).to have_css('p', text: "#{post.text.split.take(18).join(' ')}...")
+        end
+        expect(page).to have_css('.count', text: "Comments: #{post.comments.count}")
+        expect(page).to have_css('.count', text: "Likes: #{post.likes.count}")
+      end
+    end
```
Comment (jerryowusu) [inline spec/views/users/users_show_spec.rb:33]: As requested by the previous reviewer, you are missing some of the tests in this file. Kindly check the underlined tests are added and make sure they are all passing before submitting for review. :pray: 

Screenshot from 2023-09-02 00-40-16

**Review by jerryowusu — CHANGES_REQUESTED** [review]
Hi @Jrmcuervo 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

so-close

 ## Highlights
Used gitflow :+1: 
Made some requested changes :+1: 

 ## Required Changes ♻️

- [ ] As shown below, the newly migrated files are causing undefinedObject errors and duplicateObject errors that is preventing the database to be set up properly. Kindly check and fix all this issues. 

Screenshot from 2023-09-02 01-04-21

Screenshot from 2023-09-02 01-04-30



Screenshot from 2023-09-02 01-04-12



_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/users/users_show_spec.rb (hunk 4 above)
Comment (Jrmcuervo) [reply spec/views/users/users_show_spec.rb:33]: Hello! @jerryowusu  Thanks for the feedback, if you could review the show_spec.rb file in the spec/users/show_spec.rb path, there are the rest of the tests, if you do the test you will see that there are 33 tests being performed, now I fix the problem with migrations, thanks and greetings!

By the way those tests were there from the beginning @Qoosim

**Review by jerryowusu — APPROVED** [review]
Hi @Jrmcuervo 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

CookedDelectableDolphin-size_restricted


## Highlights
All tests are passing :+1: 
Great effort in making the requested changes :+1: 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**